src/app/rosetta/lib/account.ml

```diff
@@ -28,13 +30,236 @@ module Get_balance =
+  (SELECT block_id, balance from relevant_internal_block_balances)
+  UNION
+  (SELECT block_id, balance from relevant_user_block_balances)
+)
+
+SELECT
+  chain.height AS block_height,
+  balance
+FROM
+chain
+JOIN relevant_block_balances ribb ON chain.id = ribb.block_id
+WHERE chain.height <= $2
+ORDER BY (chain.height, -timestamp) DESC
+LIMIT 1
+      |}
+
+    let run (module Conn : Caqti_async.CONNECTION) start_block_height address =
+      Conn.find_opt query (address, start_block_height)
+  end
+
+  let run (module Conn : Caqti_async.CONNECTION) block_query address =
+    let open Deferred.Result.Let_syntax in
+    let%bind timing_info_opt =
+      Archive_lib.Processor.Timing_info.find_by_pk_opt
+        (module Conn)
+        (Signature_lib.Public_key.Compressed.of_base58_check_exn address)
+      |> Errors.Lift.sql ~context:"Finding timing info"
+    in
+    (* First find the block referenced by the block identifier. Then find the latest block no later than it that has a
+     * user or internal command relevant to the address we're checking and pull the balance from it. For non-vesting
+     * accounts that balance will still be the balance at the block identifier. For vesting accounts we'll also compute
+     * how much extra balance has accumulated in between the blocks. *)
+    let%bind start_block_height, start_block_hash =
+      match%bind
+        Sql.Block.run (module Conn) block_query
+        |> Errors.Lift.sql ~context:"Finding specified block"
+      with
+      | None ->
+          Deferred.Result.fail (Errors.create `Block_missing)
+      | Some (_block_id, block_info, _) ->
+          Deferred.Result.return (block_info.height, block_info.state_hash)
+    in
+    let start_block_identifier =
+      { Rosetta_models.Block_identifier.index= start_block_height
+      ; hash= start_block_hash }
+    in
+    let%bind last_relevant_command_info_opt =
+      Balance_from_last_relevant_command.run
+        (module Conn)
+        start_block_height address
+      |> Errors.Lift.sql
+           ~context:
+             "Finding balance at last relevant internal or user command."
+    in
+    let%bind computed_balance =
+      match (last_relevant_command_info_opt, timing_info_opt) with
+      | None, None ->
+          (* We've never heard of this account, at least as of the block_identifier provided *)
+          (* TODO: Should this error instead? *)
+          Deferred.Result.return 0L
+      | Some last_relevant_command_info, None ->
+          (* This account has no special vesting, so just use its last known balance *)
+          Deferred.Result.return last_relevant_command_info.balance
+      | None, Some timing_info ->
+          (* This account hasn't seen any transactions but was in the genesis ledger, so compute its balance at the start block *)
+          let open Unsigned in
+          let start_slot = UInt32.of_int 0 in
+          let end_slot = UInt32.of_int (Int.of_int64_exn start_block_height) in
```

Comment:
This should be `global_slot_since_genesis` of the block

Reply:
Thanks!